Fix cleanup on deployment error

OndroMih opened the pull request:
Before this, JNDI bindings were not cleaned up for all EJBs, leaking information to future deployments of the same name. If an app didn't deploy successfully and was deployed later, JNDI bindings remained from the previous failed deployment instead of being initialized from scratch. I saw an issue that EntityManagerFactory was closed because of a failed deployment, and then it was reused for the next deployment, leading to a closed EMF assigned to a newly deployed app. 

The cause for this was that the JNDI bindings were removed only after they were unregistered the same number of times as they were registered. During failed deployment, they were not unregistered in some cases, and an old JNDI reference was kept in the map. The fix removes all the bindings once again during application unregistration.

pzygielo commented on appserver/itest-tools/src/main/java/org/glassfish/main/itest/tools/asadmin/AsadminResultMatcher.java:
```diff
@@ -13,14 +13,13 @@
-
 /**
- * Matcher checking that {@link Asadmin} command succeeded. Prints it's output otherwise.
+ * Matcher checking that {@link Asadmin} command succeeded. Prints it's output
```
If this is changed, perhaps `it's` should become `its`.

pzygielo commented on appserver/itest-tools/src/main/java/org/glassfish/main/itest/tools/asadmin/AsadminResultMatcher.java:
```diff
@@ -30,17 +29,41 @@ private AsadminResultMatcher() {
     /**
-     * @return matcher checking that {@link Asadmin} command succeeded. Prints it's output otherwise.
+     * @return matcher checking that {@link Asadmin} command succeeded. Prints
+     * it's output otherwise.
```
as above

OndroMih committed:
Fix cleanup on deployment error

Before this, JNDI bindings were not cleaned up for all EJBs, leaking information to future deployments of the same name.
If an app didn't deploy successfully and was deployed later, JNDI bindings remained from the previous failed deployment instead of being initialized from scratch. I saw an issue that EntityManagerFactory was closed because of a failed deployment, and then it was reused for the next deployment, leading to a closed EMF assigned to a newly deployed app. 

The cause for this was that the JNDI bindings were removed only after they were unregistered the same number of times as they were registered. During failed deployment, they were not unregistered in some cases, and an old JNDI reference was kept in the map. The fix removes all the bindings once again during application unregistration.

OndroMih committed:
Fix cleanup on deployment error - renaming a variable

OndroMih committed:
Fix cleanup on deployment error - tests

OndroMih committed:
Fix cleanup on deployment error - fixes after review

OndroMih committed:
Fix cleanup on deployment error - use embedded DB in the test
